synapseclient/table.py

```diff
@@ -866,9 +869,142 @@ def _filter_duplicate_columns(self, syn, columns_to_add):
+    :param addAnnotationColumns:            If true, adds columns for all annotation keys defined across all Entities in
+                                            the SubmissionViewSchema's scope. Defaults to True.
+                                            The annotation columns will be added after a call to
+                                            :py:meth:`synapseclient.Synapse.store`.
+    :param ignoredAnnotationColumnNames:    A list of strings representing annotation names.
+                                            When addAnnotationColumns is True, the names in this list will not be
+                                            automatically added as columns to the SubmissionViewSchema if they exist in
+                                            any of the defined scopes.
+    :param properties:                      A map of Synapse properties
+    :param annotations:                     A map of user defined annotations
+    :param local_state:                     Internal use only
+
+    Example::
+        from synapseclient import SubmissionViewSchema
+
+        project_or_folder = syn.get("syn123")
+        schema = syn.store(SubmissionViewSchema(name='My Submission View', parent=project, scopes=['9614543']))
+    """
+
+    _synapse_entity_type = 'org.sagebionetworks.repo.model.table.SubmissionView'
+    _property_keys = SchemaBase._property_keys + ['scopeIds']
+    _local_keys = SchemaBase._local_keys + ['addDefaultViewColumns', 'addAnnotationColumns',
+                                            'ignoredAnnotationColumnNames']
+
+    def __init__(self, name=None, columns=None, parent=None, scopes=None,
+                 addDefaultViewColumns=True, addAnnotationColumns=True,
+                 ignoredAnnotationColumnNames=[],
+                 properties=None, annotations=None, local_state=None, **kwargs):
+
+        self.ignoredAnnotationColumnNames = set(ignoredAnnotationColumnNames)
+        super(SubmissionViewSchema, self).__init__(
+            name=name, columns=columns, properties=properties,
+            annotations=annotations, local_state=local_state, parent=parent,
+            **kwargs
+        )
+        # This is a hacky solution to make sure we don't try to add columns to schemas that we retrieve from synapse
+        is_from_normal_constructor = not (properties or local_state)
+        # allowing annotations because user might want to update annotations all at once
+        self.addDefaultViewColumns = addDefaultViewColumns and is_from_normal_constructor
+        self.addAnnotationColumns = addAnnotationColumns and is_from_normal_constructor
+
+        if self.get('scopeIds') is None:
+            self.scopeIds = []
+
+        # add the scopes last so that we can append the passed in scopes to those defined in properties
+        if scopes is not None:
+            self.add_scope(scopes)
+
+    def add_scope(self, entities):
```

Comment:
There is a number of functions that repeat in `SubmissionViewSchema` that appear in `EntityViewSchema`. Should there be a `ViewBase` class?

Reply:
They seem pretty much identical other than a few constants and how the EntityView deals with the entity types/viewmask. If we weren’t dealing with a potentially breaking change if we modified the current EntityViewSchema I I would suggest they don’t even need to be separate classes if properly parameterized. But a common base or maybe just a function that spits out an appropriately parameterized class instead of the duplication would be good here.

Reply:
I think I prefer to have `EntityViewSchema` and `SubmissionViewSchema` just because they are technically two different types.  Submissions aren't entities.  I will create a `ViewBase` class.